### PR TITLE
Porting Narrator announcement for StringCollectionEditor items

### LINKED ISSUE #2188: Port Narrator announcement for StringCollectionEditor items
* .NET Core Version: 3.1
* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**
There is no screen reader announcement for item navigation and selecting in StringCollectionEditor UI.

**Actual behavior:**
Narrator does not announce the StringCollectionEditor item when navigating up and down thru the items list.

**Expected behavior:**
Narrator announces the item when focus goes to item when navigating using keyboard/mouse.

**Minimal repro:**
* Create WinForms app with ComboBox or another list control which collection can be edited using StringCollectionEditor.
* Add PropertyGrid control to the form and associate its selected object with previously added ComboBox.
* Run the app try to edit the collection of items (add few items and navigate up/down using keyboard within StringCollectionEditor textbox.
* Observe no announcement when navigation.

### PR DESCRIPTION
**Pull request by M-Lipin**
Fixes #2188 

## Proposed changes

- Adding StringCollectionEditorTextBox class derived from standard TextBox with enhanced accessibility (UI Automation and UI Notification) support.
- Adding UIA notification announcement for item navigation within StringCollectionEditorTextBox.

## Customer Impact

- Visually impaired user will be able to distinguish the currently edited/selected/navigated string item within StringCollectionEditor control.

## Regression? 

- No

## Risk

- Minimal

### Before

No announcement for SringCollectionEditor navigation.

### After

SringCollectionEditor item navigation is announced properly.

## Test methodology 

- Manual testing.
- Automation tests (to run)
- Unit tests (to be implemented)

## Accessibility testing  

There is no UI changes but changes in accessibility providers: IAccessible is partially changed to UI Automation.
 

## Test environment(s) 

Net Core version: 3.1.0-preview1.19458.7
Microsoft Windows [Version 10.0.18362.418]

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2189)

### CODE REVIEW

File: src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs [hunk 1]
```diff
@@ -208,27 +209,255 @@ private void OKButton_click(object sender, EventArgs e)
             }
 
             // <summary>
-            //  This is called when the value property in the CollectionForm has changed.
-            //  In it you should update your user interface to reflect the current value.
+            //      This is called when the value property in the CollectionForm has changed.
+            //      In it you should update your user interface to reflect the current value.
```
Comment (RussKie) [inline src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs]: Why has this changed?

File: src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs [hunk 2]
```diff
@@ -208,27 +209,255 @@ private void OKButton_click(object sender, EventArgs e)
...
             // </summary>
+            // </doc>
```
Comment (RussKie) [inline src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs]: Please remove

File: src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs [hunk 3]
```diff
@@ -208,27 +209,255 @@ private void OKButton_click(object sender, EventArgs e)
...
             protected override void OnEditValueChanged()
             {
-                object[] items = Items;
-                string text = string.Empty;
+                _textEntry.Text = string.Join("\r\n", Items);
+            }
+
+            public class TextBoxUiaProvider : IRawElementProviderSimple
```
Comment (RussKie) [inline src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs]: Own file please

File: src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs [hunk 4]
```diff
@@ -208,27 +209,255 @@ private void OKButton_click(object sender, EventArgs e)
...
+                        // The UIA Notification event is not available, so don't attempt to raise it again.
+                        notificationEventAvailable = false;
+                    }
+
+                    return result == NativeMethods.S_OK;
+                }
+            }
 
-                for (int i = 0; i < items.Length; i++)
+            private sealed class StringCollectionEditorTextBox : TextBox
```
Comment (RussKie) [inline src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/StringCollectionEditor.cs]: Own file please

File: src/System.Windows.Forms.Design.Editors/src/Misc/NativeMethods.cs [hunk 5]
```diff
@@ -991,12 +991,119 @@ public override string ToString()
...
-        //SendMessage(IntPtr, int, bool, IntPtr), SendMessage(IntPtr, int, IntPtr, ListViewCompareCallback),
-        //SendMessageW, SendMessageA, ValidateRect(IntPtr, ref RECT), ValidateRgn(IntPtr, IntPtr)
-        //COMRECT.FromXYWH, RECT.FromXYWH
-
         public const int MAX_PATH = 260;
+
+        public const int UiaRootObjectId = -25;
+
+        // UI Automation PropertyIds
+        internal const int UIA_RuntimeIdPropertyId = 30000;
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design.Editors/src/Misc/NativeMethods.cs]: I really feel we should separate out these UIA consts to an enum

**Comment by RussKie** [comment]
Please rebase on top of release/3.1-uitypeeditors, force-push and reopen

**Pull request closed without merging**